tianshou/data/collector.py

```diff
@@ -220,198 +298,299 @@ def collect(
-                break
+                if n_episode:
+                    if sample_equal_num_episodes_per_worker:
+                        non_idle_env_ids = (
+                            self.sample_equal_episodes_per_worker_postprocessing_on_done_env(
+                                gym_reset_kwargs,
+                                non_idle_env_ids,
+                                np.where(~is_done_and_non_idle)[0],
+                            )
+                        )
+                    else:
+                        non_idle_env_ids = (
+                            self.sample_at_least_one_episode_per_worker_postprocessing_on_done_env(
+                                episode_count,
+                                env_ids_done,
+                                n_episode,
+                                non_idle_env_ids,
+                            )
+                        )
+                    # todo check if async collector is equivalent to just passing here
 
-        # generate statistics
-        self.collect_step += step_count
-        self.collect_episode += episode_count
-        collect_time = max(time.time() - start_time, 1e-9)
-        self.collect_time += collect_time
+            self.data.obs = self.data.obs_next
 
+        collect_call_duration = max(time.time() - start_time, 1e-9)
+        self._update_collection_counters(collect_call_duration, episode_count, step_count)
         if n_episode:
-            data = Batch(
-                obs={},
-                act={},
-                rew={},
-                terminated={},
-                truncated={},
-                done={},
-                obs_next={},
-                info={},
-                policy={},
-            )
-            self.data = cast(RolloutBatchProtocol, data)
+            # like self.reset but without resetting the collection counters
+            self._reset_data()
             self.reset_env()
+        return CollectStats.from_collect_output(
+            step_count=step_count,
+            episode_count=episode_count,
+            collect_call_duration=collect_call_duration,
+            episode_returns=np.array(episode_returns),
+            episode_lens=np.array(episode_lens, int),
+        )
 
-        return CollectStats(
-            n_collected_episodes=episode_count,
-            n_collected_steps=step_count,
-            collect_time=collect_time,
-            collect_speed=step_count / collect_time,
-            returns=np.array(episode_returns),
-            returns_stat=SequenceSummaryStats.from_sequence(episode_returns)
-            if len(episode_returns) > 0
-            else None,
-            lens=np.array(episode_lens, int),
-            lens_stat=SequenceSummaryStats.from_sequence(episode_lens)
-            if len(episode_lens) > 0
-            else None,
+    def sample_at_least_one_episode_per_worker_postprocessing_on_done_env(
```

Comment:
this doesn't seem to sample anything, rather only filter.

Marking for discussion